### ISSUE TITLE
sys-cluster/pcs help needed

### ISSUE THREAD

**Issue by Alessandro-Barbieri** [gentoo/guru#24]
@custom82 is trying to package pcs but we have divergent ideas on how to package it. By default it will install ruby stuff in `/usr/lib/pcs` but I think it should use the ruby eclass to handle ruby stuff properly.

**Comment by telans** [comment]
`ruby-single` would be suited for this right?

**Comment by custom82** [comment]
it is not a ruby ​​library, it is code belonging to a daemon

**Comment by AndrewAmmerlaan** [comment]
> it is not a ruby ​​library, it is code belonging to a daemon

In that case I don't think it is necessary to install into ruby specific directories. However, you'll still need the ruby-* eclass to correctly depend on ruby packages. This is to ensure that all required ruby packages are installed with correct RUBY_TARGETS, otherwise package X might be installed for ruby25 and package Y with ruby26, resulting in that your package will not work.

**Comment by custom82** [comment]
for this there is the useflag RUBY_TARGETS to make sure that a library is installed on the available targets

**Comment by custom82** [comment]
pcs has no need for eclass ruby-ng

**Comment by AndrewAmmerlaan** [comment]
> for this there is the useflag RUBY_TARGETS to make sure that a library is installed on the available targets

Yes but the ebuild still needs to enforce this with the `ruby_add_rdepend` function, similar to how we need `[${PYTHON_USEDEP}]` to pull in python packages with the correct PYTHON_TARGETS. There's no guarantee that a user does not override the default RUBY_TARGETS on a per package basis in package.use, thus the ebuild has to explicitly depend on the same ruby_target for all ruby packages.

**Comment by AndrewAmmerlaan** [comment]
> pcs has no need for eclass ruby-ng

We do need some ruby eclass for the `ruby_add_rdepend` function, whether ruby-ng is the best on to use in this case I'm not sure to be honest.

**Comment by custom82** [comment]
the ruby_add_rdepend function is declared in the ruby-ng and ruby-fakegem eclass

**Comment by AndrewAmmerlaan** [comment]
> the ruby_add_rdepend function is declared in the ruby-ng and ruby-fakegem eclass

Then we'll need one of those two. I lack the experience with ruby necessary to know which one we'll need. But I see that @Alessandro-Barbieri has a ruby package in ::guru so perhaps he knows.

**Comment by AndrewAmmerlaan** [comment]
ruby-fakegems inherits ruby-ng:

```
# @ECLASS: ruby-fakegem.eclass
# @MAINTAINER:
# Ruby herd <ruby@gentoo.org>
# @AUTHOR:
# Author: Diego E. Pettenò <flameeyes@gentoo.org>
# Author: Alex Legler <a3li@gentoo.org>
# Author: Hans de Graaff <graaff@gentoo.org>
# @SUPPORTED_EAPIS: 4 5 6 7
# @BLURB: An eclass for installing Ruby packages to behave like RubyGems.
# @DESCRIPTION:
# This eclass allows to install arbitrary Ruby libraries (including Gems),
# providing integration into the RubyGems system even for "regular" packages.

inherit ruby-ng
```

Therefore ruby-ng is the smaller eclass so my guess would be that this is the correct one for this use case if we do not need the extra stuff provided by the ruby-fakegem eclass.

**Comment by custom82** [comment]
do you think this works for the case?
ruby_add_rdepend() {
        case $# in
                1) ;;
                2)
                        case ${EAPI} in
                                4|5|6)
                                        [[ "${GENTOO_DEV}" == "yes" ]] && eqawarn "You can now use the usual syntax in ruby_add_rdepend for $CATEGORY/$PF"
                                        ruby_add_rdepend "$(_ruby_wrap_conditions "$1" "$2")"
                                        return
                                        ;;
                                *)
                                        die "Use the usual depend syntax with a single argument in ruby_add_rdepend"
                                        ;;
                        esac
                        ;;
                *)
                        die "bad number of arguments to $0"
                        ;;
        esac

        local dependency=$(_ruby_atoms_samelib "$1")

        RDEPEND="${RDEPEND} $dependency"

        # Add the dependency as a test-dependency since we're going to
        # execute the code during test phase.
        case ${EAPI} in
                4|5|6) DEPEND="${DEPEND} test? ( ${dependency} )" ;;
                *) BDEPEND="${BDEPEND} test? ( ${dependency} )" ;;
        esac
        if ! has test "$IUSE"; then
                IUSE+=" test"
                RESTRICT+=" !test? ( test )"
        fi
}

**Comment by custom82** [comment]
ruby-ng and ruby-fakegem are for building ruby ​​libraries not for the purpose

**Comment by AndrewAmmerlaan** [comment]
> ruby-ng and ruby-fakegem are for building ruby ​​libraries not for the purpose

Regardless of whether we are building a library or not, we do need some way to take care of the ruby_targets_foobar use flags on the ruby dependencies. The `ruby_add_rdepend` does just that, and I don't see any other way of doing it with e.g. the ruby-utils or ruby-single eclasses, but I might be missing something.

The ebuild is working correctly at the moment, is it not?

**Comment by custom82** [comment]
no

**Comment by AndrewAmmerlaan** [comment]
> no

:O why not?

With the `ruby_add_rdepend` function we get this in RDEPEND:

```
* 
*      media-libs/fontconfig
*      dev-libs/libffi
*      sys-apps/coreutils
* 
*      >=www-servers/tornado-6.0
*      <www-servers/tornado-7.0
*      dev-python/python-dateutil[python_targets_python3_7(-)?,python_targets_python3_8(-)?,python_targets_python3_9(-)?,-python_single_target_python3_7(-),-python_single_target_python3_8(-),-python_single_target_python3_9(-)]
*      dev-python/distro[python_targets_python3_7(-)?,python_targets_python3_8(-)?,python_targets_python3_9(-)?,-python_single_target_python3_7(-),-python_single_target_python3_8(-),-python_single_target_python3_9(-)]
*      dev-python/dacite[python_targets_python3_7(-)?,python_targets_python3_8(-)?,python_targets_python3_9(-)?,-python_single_target_python3_7(-),-python_single_target_python3_8(-),-python_single_target_python3_9(-)]
*      dev-python/lxml[python_targets_python3_7(-)?,python_targets_python3_8(-)?,python_targets_python3_9(-)?,-python_single_target_python3_7(-),-python_single_target_python3_8(-),-python_single_target_python3_9(-)]
*      dev-python/pyopenssl[python_targets_python3_7(-)?,python_targets_python3_8(-)?,python_targets_python3_9(-)?,-python_single_target_python3_7(-),-python_single_target_python3_8(-),-python_single_target_python3_9(-)]
*      dev-python/pycurl[python_targets_python3_7(-)?,python_targets_python3_8(-)?,python_targets_python3_9(-)?,-python_single_target_python3_7(-),-python_single_target_python3_8(-),-python_single_target_python3_9(-)]
*      dev-python/pyparsing[python_targets_python3_7(-)?,python_targets_python3_8(-)?,python_targets_python3_9(-)?,-python_single_target_python3_7(-),-python_single_target_python3_8(-),-python_single_target_python3_9(-)]
*      sys-process/psmisc
*      dev-libs/openssl
*      sys-libs/pam
*      >=sys-cluster/corosync-3.0
*      >=sys-cluster/pacemaker-2.0
*  ruby_targets_ruby25? (
* dev-ruby/bundler[ruby_targets_ruby25(-)]
* dev-ruby/rubygems[ruby_targets_ruby25(-)]
* dev-ruby/open4[ruby_targets_ruby25(-)]
* dev-ruby/highline[ruby_targets_ruby25(-)]
* dev-ruby/json[ruby_targets_ruby25(-)]
* dev-ruby/multi_json[ruby_targets_ruby25(-)]
* dev-ruby/rack[ruby_targets_ruby25(-)]
* dev-ruby/rack-protection[ruby_targets_ruby25(-)]
* dev-ruby/rack-test[ruby_targets_ruby25(-)]
* dev-ruby/thor[ruby_targets_ruby25(-)]
* dev-ruby/ethon[ruby_targets_ruby25(-)]
* dev-ruby/thin[ruby_targets_ruby25(-)]
* dev-ruby/tilt[ruby_targets_ruby25(-)]
* dev-ruby/sinatra[ruby_targets_ruby25(-)]
* dev-ruby/open4[ruby_targets_ruby25(-)]
* dev-ruby/backports[ruby_targets_ruby25(-)]
* )
* ruby_targets_ruby26? (
* dev-ruby/bundler[ruby_targets_ruby26(-)]
* dev-ruby/rubygems[ruby_targets_ruby26(-)]
* dev-ruby/open4[ruby_targets_ruby26(-)]
* dev-ruby/highline[ruby_targets_ruby26(-)]
* dev-ruby/json[ruby_targets_ruby26(-)]
* dev-ruby/multi_json[ruby_targets_ruby26(-)]
* dev-ruby/rack[ruby_targets_ruby26(-)]
* dev-ruby/rack-protection[ruby_targets_ruby26(-)]
* dev-ruby/rack-test[ruby_targets_ruby26(-)]
* dev-ruby/thor[ruby_targets_ruby26(-)]
* dev-ruby/ethon[ruby_targets_ruby26(-)]
* dev-ruby/thin[ruby_targets_ruby26(-)]
* dev-ruby/tilt[ruby_targets_ruby26(-)]
* dev-ruby/sinatra[ruby_targets_ruby26(-)]
* dev-ruby/open4[ruby_targets_ruby26(-)]
* dev-ruby/backports[ruby_targets_ruby26(-)]
* )
```

which is how it should be, right?

**Comment by custom82** [comment]
because eclass ruby-ng changes the source directory

**Comment by custom82** [comment]
forget the ruby27 version?

**Comment by AndrewAmmerlaan** [comment]
> forget the ruby27 version?

No, dependency dev-ruby/thin does not support ruby27, because it's dependency dev-ruby/eventmachine does not support ruby27:
```
eix eventmachine
* dev-ruby/eventmachine
Available versions:  1.2.7^t{xpak} {doc test RUBY_TARGETS="ruby25 ruby26"}
Homepage:            https://github.com/eventmachine/eventmachine
Description:         EventMachine is a fast, simple event-processing library for Ruby programs
```
And that is exactly why we need to use the `ruby_add_rdepend` function, without it portage would install this package with ruby27 and it will fail because the dependency of the dependency does not support ruby27.

> because eclass ruby-ng changes the source directory

Alright, we can change that back by setting the S variable if needed, but I emerged this yesterday and it seemed fine.

**Comment by custom82** [comment]
I don't know who did the ebuild but here it says it's supported!

https://github.com/eventmachine/eventmachine
EventMachine supports Ruby 2.0.0 through 2.7, JRuby and works well on Windows as well as many operating systems from the Unix family (Linux, Mac OS X, BSD flavors).

**Comment by AndrewAmmerlaan** [comment]
> I don't know who did the ebuild but here it says it's supported!
> 
> https://github.com/eventmachine/eventmachine
> EventMachine supports Ruby 2.0.0 through 2.7, JRuby and works well on Windows as well as many operating systems from the Unix family (Linux, Mac OS X, BSD flavors).

In that case we could make a PR to ::gentoo to add ruby27 to RUBY_USE in that ebuild.

**Comment by custom82** [comment]
* pcs-0.10.7.tar.gz BLAKE2B SHA512 size ;-) ...                                                                                                                                                                                     [ ok ]
>>> Unpacking source...
>>> Unpacking pcs-0.10.7.tar.gz to /var/tmp/portage/sys-cluster/pcs-0.10.7/work/all
>>> Source unpacked in /var/tmp/portage/sys-cluster/pcs-0.10.7/work
>>> Preparing source in /var/tmp/portage/sys-cluster/pcs-0.10.7/work ...
 * Running prepare phase for all ...
 * Applying remove-ruby-bundle-path.patch ...
patching file pcs/settings_default.py
Hunk #1 succeeded at 91 with fuzz 1 (offset 10 lines).                                                                                                                                                                               [ ok ]
 * Applying openrc-0.10.7.patch ...
patching file pcs/settings_default.py
patching file pcsd/pcs.rb
Hunk #1 succeeded at 1709 (offset -3 lines).
Hunk #2 succeeded at 1722 with fuzz 2.
Hunk #3 succeeded at 1775 (offset -3 lines).
Hunk #4 succeeded at 1803 (offset 37 lines).
Hunk #5 succeeded at 1818 (offset 37 lines).
patching file pcs/lib/external.py
Hunk #1 succeeded at 192 with fuzz 2 (offset 123 lines).
Hunk #2 succeeded at 214 with fuzz 2 (offset 121 lines).
Hunk #3 succeeded at 295 (offset 118 lines).
Hunk #4 succeeded at 348 (offset 119 lines).                                                                                                                                                                                         [ ok ]
 * Running prepare phase for all ...
 * Running source copy phase for ruby27 ...
>>> Source prepared.
>>> Configuring source in /var/tmp/portage/sys-cluster/pcs-0.10.7/work ...
>>> Source configured.
>>> Compiling source in /var/tmp/portage/sys-cluster/pcs-0.10.7/work ...
>>> Source compiled.
 * Skipping make test/check due to ebuild restriction.
>>> Test phase [disabled because of RESTRICT=test]: sys-cluster/pcs-0.10.7

>>> Install sys-cluster/pcs-0.10.7 into /var/tmp/portage/sys-cluster/pcs-0.10.7/image
make -j20 install SYSTEMCTL_OVERRIDE= DESTDIR=/var/tmp/portage/sys-cluster/pcs-0.10.7/image CONF_DIR=/etc/default/ PREFIX=/usr BUNDLE_INSTALL_PYAGENTX=false BUNDLE_TO_INSTALL=false BUILD_GEMS=false 
make: *** No rule to make target 'install'.  Stop.
 * ERROR: sys-cluster/pcs-0.10.7::guru failed (install phase):
 *   emake failed
 * 
 * If you need support, post the output of `emerge --info '=sys-cluster/pcs-0.10.7::guru'`,
 * the complete build log and the output of `emerge -pqv '=sys-cluster/pcs-0.10.7::guru'`.
 * The complete build log is located at '/var/tmp/portage/sys-cluster/pcs-0.10.7/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/sys-cluster/pcs-0.10.7/temp/environment'.
 * Working directory: '/var/tmp/portage/sys-cluster/pcs-0.10.7/work'
 * S: '/var/tmp/portage/sys-cluster/pcs-0.10.7/work'

 * Messages for package sys-cluster/pcs-0.10.7:

 * ERROR: sys-cluster/pcs-0.10.7::guru failed (install phase):
 *   emake failed
 * 
 * If you need support, post the output of `emerge --info '=sys-cluster/pcs-0.10.7::guru'`,
 * the complete build log and the output of `emerge -pqv '=sys-cluster/pcs-0.10.7::guru'`.
 * The complete build log is located at '/var/tmp/portage/sys-cluster/pcs-0.10.7/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/sys-cluster/pcs-0.10.7/temp/environment'.
 * Working directory: '/var/tmp/portage/sys-cluster/pcs-0.10.7/work'
 * S: '/var/tmp/portage/sys-cluster/pcs-0.10.7/work'

**Comment by custom82** [comment]
thunderdome /var/db/repos/guru/sys-cluster/pcs # ls /var/tmp/portage/sys-cluster/pcs-0.10.7/work/
all  ruby27

**Comment by AndrewAmmerlaan** [comment]
Fixed in: https://gitweb.gentoo.org/repo/proj/guru.git/commit/?id=24883715f6fab5df1a2d8cf7c7ca5830c5d588a2

**Comment by custom82** [comment]
the ebuild in the dev branch works

**Comment by custom82** [comment]
now we need to fix the patch for openrc

**Comment by Alessandro-Barbieri** [comment]
@AndrewAmmerlaan I have a ruby package in ::guru and it's another story because it's a ruby gem

**Comment by Alessandro-Barbieri** [comment]
@custom82 please use markdown backticks to format code, you have mentioned all the guru issues

**Comment by Alessandro-Barbieri** [comment]
We can't let make install the python module, we have to properly use the eclass
